Various fixes for TOML documentation

DavidFair opened the pull request:
**Description of work.**
- Add section on toml_file_format
- Add section on [metadata] and [instrument]
- Correct detector.calibration to detector.correction that got missed

**Report to:** ISIS SANS/Rob

**To test:**
- Check doc tests pass
- Review new docs

*There is no associated issue.*

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Various fixes for TOML documentation

- Add section on toml_file_format
- Add section on [metadata] and [instrument]
- Correct detector.calibration to detector.correction that got missed

smk78 committed:
WIP: Overhauling command explanations

smk78 committed:
Oops, fix broken hyperlink (used wrong markup!)

DavidFair committed:
fixup! WIP: Overhauling command explanations

DavidFair commented:
Thanks Steve, I've run pre-commit over your changes to fix them up

DavidFair commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -412,49 +546,144 @@ DET/OVERLAP x y
-**Replacement**
+Times were specified in microseconds.
+
+# David this command used to be used on LOQ. Does the note on
```
So this predates me, so I went on the implementation

`FIT /CENTRE` is [ignored by the old parser](https://github.com/mantidproject/mantid/blob/2dcb0847d0c0250501c677c88dc7beee995ba662/scripts/SANS/sans/user_file/user_file_parser.py#L2366)
whilst 
`FIT /MONITOR` actually [was parsed](https://github.com/mantidproject/mantid/blob/main/scripts/SANS/sans/user_file/user_file_parser.py#L1766) and was used in [prompt peak corrections](https://github.com/mantidproject/mantid/blob/2dcb0847d0c0250501c677c88dc7beee995ba662/scripts/SANS/sans/user_file/txt_parsers/ParsedDictConverter.py#L1083) and AFAIK can be used on anything (at least there are no checks where it's [used](https://github.com/mantidproject/mantid/blob/main/scripts/SANS/sans/algorithm_detail/calculate_sans_transmission.py#L293))

smk78 replied:
Ok, I'll edit the file accordingly.

DavidFair reviewed (commented):
Thanks for going through in detail on what the behaviour of each command does. It greatly improves the docs!

I've left some feedback + answers to questions in the form of reviews
I've also pushed some doc fixes directly in answer to your questions, changed `shift.distance` to `shift.factor` and replaced `See Also: foo` with auto generated links

DavidFair committed:
Various TOML docs tidying

- Fix links to use autogeneration
- Answer some ambigious points
- Address in-line comments
- Rename shift.distance to shift.factor

DavidFair committed:
Change shift.distance to shift.factor

Since the shift was a factor, not a distance as I incorrectly assumed.

This should be safe as we're still in V0 of the TOML format to find
problems like this

DavidFair commented:
I've marked this as a draft as we work through the rest of the docs, feel free to push at any point as I'll hold off touching this branch again until your changes are ready

smk78 committed:
More overhauling of command explainations

Commiting Steve's work on top of the fixes I've already done to help
keep history nice and clean.

smk78 commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -142,38 +230,61 @@ Note: This command subtracts the specified background level from the specified m
+*increases the apparent radius from the rotation axis of the detector
+to the active plane*.
+
+# David how are the X/Y/ZTILT commands defined? What values are they modifying?
```
Ok!
But the Replacement Example shows
```
    front_x_tilt = -0.000085
    front_y_tilt = 0.0001419
```
Is that correct; it's not front_xtilt and front_ytilt, for example, as per the Existing Example?

DavidFair replied:
I had to double check, it's `front_x_tilt` as per: https://github.com/mantidproject/mantid/blob/2fa08eec7a58a91e3047424d9491d1b824530f63/scripts/SANS/sans/user_file/toml_parsers/toml_v1_parser.py#L211

IIRC the reason I went for this is to make it consistent throughout the file using snake_case_between_all_words

smk78 replied:
👍

smk78 commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -10,7 +10,10 @@ SANS TOML Files
 
-- Lengths are given in meters within TOML files, unlike previous legacy formats.
+- Lengths are *always* specified in meters within TOML files, unlike previous legacy formats.
+- Angles are specified in degrees.
```
That's fine. Was just checking radians weren't expected anywhere!

smk78 commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -346,34 +463,46 @@ DET/SHIFT x
-------------------
+# David your existing doc says that the shift is a distance; that is
+# inaccurate so I suggest it be changed to factor as for rescale.
+# I've changed this doc accordingly.
```
👍

smk78 committed:
Resolving feedback on comment

smk78 commented:
Comment on MASK/LINE needs a response.

smk78 commented:
Comment on MON/TRANS pending further discussion.

smk78 commented:
Is comment on QRESOL[/H1=x1][/W1=y1][/H2=x2][/W2=y2] resolved?

smk78 commented:
Comment on SAMPLE/PATH[/ON][/OFF] needs a response.

smk78 commented:
Comment on SET CENTRE[/MAIN][/HAB] a b [c d] needs a response.

smk78 commented:
Comments on SET SCALES a b c d e need a response.

smk78 commented:
Comment on TRANS/TRANSPEC=s/SHIFT=dz pending further discussion.

DavidFair commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -1152,128 +1820,247 @@ in :ref:`algm-TOFSANSResolutionByPixel`.
+
+.. ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
+.. David you had these as mm even in TOML; I assume that was a mistake?
+.. ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
```
Yeah, everything is in meters so that mistakes like this becomes obvious. Thanks for fixing thank 👍

DavidFair commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -1152,128 +1820,247 @@ in :ref:`algm-TOFSANSResolutionByPixel`.
+
+Unsupported
+
+.. ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
```
In this case, it's supported in the existing user files but not TOML as we haven't selected a name / location for this.

@smk78 If you could send me a user file which uses it (to write a test to make sure we get the same result from the old / new files), and we could decide what to call it / where to put it?
Internally it uses an attribute called `wide_angle_correction`, so maybe this would be a good name?

smk78 replied:
Ok, I'll need to investigate.

smk78 replied:
So...

The (now unsupported) legacy commands MON/HABEFF & MON/HABPATH were in routine use on LOQ upto ~start of 2011. This date approximately corresponds to the point at which LOQ data reduction moved over from COLETTE to Mantid. The new Mantid User Files then noted:

```
PRINT MON/HABEFF correction not yet implemented
PRINT MON/HABPATH correction not yet implemented
```

It looks to me as if the (supported) legacy command SAMPLE/PATH was developed at SNS and that ISIS was unaware of it, because I cannot find an instance of it being used in a LOQ User File (and LOQ would be the instrument that would benefit most, followed by SANS2D). Thus it would be good to have this command TOML-ised, please. Maybe something like:

```
[transmission]
wide_angle_correction = true | false
```

smk78 replied:
Some observations from testing SAMPLE/PATH:

- we will need a field added to the SANS Interface to show if it is enabled or not (in Settings > Adjustment > Transmission is probably the most appropriate place)
- if the can transmission being corrected is ~unity - as can often happen if we just use the instrument background as the can and so Can Transmission and the Can Direct are both the Direct Beam - and the front/hab detector is selected then the reduced intensities and their uncertainties can get output as -Nan. Bit odd. The rear/main detector data reduce fine.
- The transmission workspaces that are written (or the values to file) appear to be identical whether or not SAMPLE/PATH is ON or not. I assume this means that the correction is applied internally after these values are generated. It would be good to be able to see the corrected values.
- This correction, unsurprisingly, has negligible impact on rear/main detector data:
![image](https://user-images.githubusercontent.com/10938679/146806016-187f360d-84ae-4d33-97c6-3d8725524b0b.png)
but it does have a small effect on the front/hab data
![image](https://user-images.githubusercontent.com/10938679/146806102-6cdbbaad-142a-4480-8fcb-f3dcac9df778.png)
Note that a different sample, for example, one that is more absorbing than this sample, will likely show a bigger effect.

[USER_LOQ_211G_M3_STEVE_TESTING_TOML__M3_main_log__DONT_USE_With_SamplePath.txt](https://github.com/mantidproject/mantid/files/7746932/USER_LOQ_211G_M3_STEVE_TESTING_TOML__M3_main_log__DONT_USE_With_SamplePath.txt)
[USER_LOQ_211G_M3_STEVE_TESTING_TOML__M3_hab_log__DONT_USE_With_SamplePath.txt](https://github.com/mantidproject/mantid/files/7746934/USER_LOQ_211G_M3_STEVE_TESTING_TOML__M3_hab_log__DONT_USE_With_SamplePath.txt)
Data are from LOQ Cycle 20/2
![image](https://user-images.githubusercontent.com/10938679/146806595-a1374fa6-45f9-426c-9574-cea2b985d12d.png)

DavidFair replied:
We'll have to push this off to a separate issue, I agree this is a good idea but I'm worried about scope creep
This PR has flagged up a number of things as we go through the docs with a fine-toothed comb

smk78 replied:
Agreed.

Have opened https://github.com/mantidproject/mantid/issues/33543 to catch the requirement.

DavidFair committed:
Address some ambigious points in TOML docs

Addresses some outstanding points raised where it's not obvious what the
behaviour or syntax required is

DavidFair commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -1284,36 +2068,67 @@ set centre a b
+    SANS2D: ( 0.100, -0.080)
+    ZOOM:   (-0.170, -0.050)
+
+**Note: In 2020 it was discovered that due to a forever bug in the legacy
```
So this command is parsed correctly by both the legacy and existing parser. It was only `SET SCALES` which had the problems...

smk78 replied:
Ok, I have a couple of questions here:

1. So you're saying if the legacy parser encountered SET CENTRE a b c d it used (a,b) as the centre coords and passed pixel sizes c & d to the data reduction, right? What did it do if c & d were _not_ specified?

2. When you say this command is also parsed by the existing parser, do you mean the TOML parser? If so, what is the replacement syntax for specifying c & d?

DavidFair replied:
> 1. So you're saying if the legacy parser encountered SET CENTRE a b c d it used (a,b) as the centre coords and passed pixel sizes c & d to the data reduction, right? What did it do if c & d were not specified?

It would throw `SetParser: Expected 5 entries for the SCALES setting, but got n` (see next link)

> 2. When you say this command is also parsed by the existing parser, do you mean the TOML parser? If so, what is the replacement syntax for specifying c & d?

Under the hood `SET SCALES` parses into `s, a, b, c, d`:
https://github.com/mantidproject/mantid/blob/b4bb8da80b0e8824dee5cec948f01f22ea948d08/scripts/SANS/sans/user_file/user_file_parser.py#L1242

However only `s` was taken (and we only have space for s in the data structure) here (notice it `scales.s`):
https://github.com/mantidproject/mantid/blob/b4bb8da80b0e8824dee5cec948f01f22ea948d08/scripts/SANS/sans/user_file/txt_parsers/ParsedDictConverter.py#L931

And the state object only has space for `s`:
https://github.com/mantidproject/mantid/blob/b4bb8da80b0e8824dee5cec948f01f22ea948d08/scripts/SANS/sans/state/StateObjects/StateScale.py#L30

Unfortunately, this pre-dates me so I can't say why it's doing what it's doing or what the correct behaviour is. So there's still no support for the 4 other values

smk78 replied:
One of us is confused here, @DavidFair ! My questions above were about SET CENTRE, but you've replied about SET SCALES!

DavidFair replied:
https://github.com/mantidproject/mantid/blob/main/scripts/SANS/sans/user_file/user_file_parser.py#L1183

```
        An undocumented feature is:
        SET CENTRE[/MAIN|/HAB] x y [d1 d2]
        where d1 and d2 are pixel sizes. This is not used in the old parser, but user files have it nonetheless.
```

You'll notice at https://github.com/mantidproject/mantid/blob/main/scripts/SANS/sans/user_file/user_file_parser.py#L1250 and it simply extracts two float values, so `d1` and `d2` (from the above docs) are ignored AFAIK

This is probably why I ended up looking at SCALES by accident instead

smk78 replied:
So how is the size of the detector pixels handled? What values does Mantid use?

DavidFair commented on docs/source/techniques/SANS-Toml-v1.rst:
```diff
@@ -1322,37 +2137,139 @@ set centre a b c d [/MAIN] [/HAB]
+User File command parser the parameters b, c, d & e have never been implemented
+in Mantid.** See this `issue <https://github.com/mantidproject/mantid/issues/27948>`_.
+
+.. ZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
```
So this bug is effectively here, where `b, c, d, and e` are completely left behind:
https://github.com/mantidproject/mantid/blob/cffb8050f43a522a2ff059a7834b0e4f59aa1075/scripts/SANS/sans/user_file/txt_parsers/ParsedDictConverter.py#L934

This means that it will fall back here https://github.com/mantidproject/mantid/blob/cffb8050f43a522a2ff059a7834b0e4f59aa1075/scripts/SANS/sans/algorithm_detail/scale_sans_workspace.py#L28
To always multiplying by `s` regardless of the workspace, or a factor of 100 if nothing was provided

smk78 replied:
Ok. This is now clear in the Docs. And the https://github.com/mantidproject/mantid/issues/27948 will need to be addressed at a convenient point.

smk78 reviewed (commented):
To MASK/LINE you have added
```
An effectively infinite cylinder (length 100m) with the angle and width set
by the user is created at x=0, y=0, and z equal to the detector depth.
```
My problem with this is the meaning of x=0, y=0. I'm presuming these are 'relative' coordinates once the beam centre has been established? But in the context of this document I don't see how these can have meaning. Would it be wrong to change 'at x=0, y=0, and z equal to the detector depth.' into 'at the beam centre coordinates with a depth equal to that of the detector (depth).'?

smk78 reviewed (commented):
In response to your email reply regarding my query about the additional TRANS command qualifiers I have added TRANS/MASK, TRANS/RADIUS and TRANS/ROI to the document.

smk78 committed:
And more overhauling of command explanations

DavidFair commented:
> To MASK/LINE you have added
> 
> ```
> An effectively infinite cylinder (length 100m) with the angle and width set
> by the user is created at x=0, y=0, and z equal to the detector depth.
> ```
> 
> My problem with this is the meaning of x=0, y=0. I'm presuming these are 'relative' coordinates once the beam centre has been established? But in the context of this document I don't see how these can have meaning. Would it be wrong to change 'at x=0, y=0, and z equal to the detector depth.' into 'at the beam centre coordinates with a depth equal to that of the detector (depth).'?

Looking and quoting from various SANS instrument IDFs' comments:
> The z-axis is set parallel to and in the direction of the beam. the 
    y-axis points up and the coordinate system is right handed

DavidFair committed:
Clarify x=0, y=0 for cylinder masking

Since it's not obvious which directions x, y and z lie in without having
to refer back to the associated IDF files.

DavidFair committed:
Bump version for these changes to 6.3

Since they didn't land in 6.2 with this PR as it's mostly docs

DavidFair committed:
Update TOML file for SystemTest

Since we have changed "distance" to "factor" to better match the
previous parser implementation

smk78 commented:
Regarding https://github.com/mantidproject/mantid/pull/33065#issuecomment-1031513016 and the new wording:
![image](https://user-images.githubusercontent.com/10938679/153496469-0148085b-c8bc-4cf8-b0c0-d2995ee752a1.png)

I think I understand now. But the new description still can't be right.

For example, because long wavelength neutrons drop under gravity the y coord of where the beam strikes the detector will not be the same as the y coord of the source (and there will probably be a slight shift in x too for instruments with a bender). So the origin for the line being drawn by this command must surely be where the beam impinges on the detector, right? Now in 'instrument speak' this location is what we call the beam centre coordinates or the centre of the pattern. My point is that the (x,y) coordinates of this position in the Mantid coordinate frame will often be different to those of the source (0,0).

So, how about something like: "...by the user is created in the plane of the detector from the point at which the transmitted beam is incident on the detector"?

smk78 reviewed (dismissed):
Of the four changed files here, the changes to the 1st, 3rd & 4th were by @DavidFair, the latter two at my request. I approve these changes.
However, the vast majority of the changes to the 2nd file (SANS-Toml-v1.rst) are mine so it is probably unethical for me to formally approve my own work (though I warrant that these documentation changes are fine)!

smk78 committed:
Text change in MASK/LINE

Text change as per https://github.com/mantidproject/mantid/pull/33065#issuecomment-1035553371

DavidFair reviewed (commented):
I've reviewed Steve's work and I'm happy with this
This PR is mostly doc changes, with some tweaks to our v0 file format to use more correct terminology
